Add doc for neural sparse search with 2 new analyzers (#10062)
opensearch-project/documentation-website: _vector-search/ai-search/neural-sparse-custom.md

---
layout: default
title: Using custom configurations for neural sparse search
parent: Neural sparse search
grand_parent: AI search
nav_order: 20
has_children: false
redirect_from:
  - /search-plugins/neural-sparse-with-pipelines/
---

# Using custom configurations for neural sparse search

Neural sparse search using automatically-generated vector embeddings operates in two modes: doc-only and bi-encoder. For more information, see [Generating sparse vector embeddings automatically]({{site.url}}{{site.baseurl}}/vector-search/ai-search/neural-sparse-with-pipelines/).

At query time, you can use custom models in the following ways: 

- **Bi-encoder mode**: Use your deployed sparse encoding model to generate embeddings from query text. This must be the same model you used at ingestion time.

- **Doc-only mode with a custom tokenizer**: Use your deployed tokenizer model to tokenize query text. The token weights are obtained from a precomputed lookup table.

The following is a complete example of using a custom model for neural sparse search.

## Step 1: Configure a sparse encoding model/tokenizer

> Both the bi-encoder and doc-only with a custom tokenizer search modes require you to configure a sparse encoding model for ingestion. Bi-encoder mode uses the same model for search; doc-only mode uses a separate tokenizer for search.

Reviewer: ```suggestion
You must configure a sparse encoding model for ingestion when using both the bi-encoder mode and the doc-only mode with a custom tokenizer. Bi-encoder mode uses the same model for search; doc-only mode uses a separate tokenizer for search.
```

#### Doc-only mode with a custom tokenizer

When using the doc-only mode with a custom tokenizer, you need to register the `amazon/neural-sparse/opensearch-neural-sparse-encoding-doc-v3-distill` model, which you'll use at ingestion time, and the `amazon/neural-sparse/opensearch-neural-sparse-tokenizer-v1` tokenizer, which you'll use at search time.

Register the sparse encoding model:

```json
POST /_plugins/_ml/models/_register?deploy=true
{
  "name": "amazon/neural-sparse/opensearch-neural-sparse-encoding-doc-v3-distill",
  "version": "1.0.0",
  "model_format": "TORCH_SCRIPT"
}
```
{% include copy-curl.html %}

Register the tokenizer:

```json
POST /_plugins/_ml/models/_register?deploy=true
{
  "name": "amazon/neural-sparse/opensearch-neural-sparse-tokenizer-v1",
  "version": "1.0.1",
  "model_format": "TORCH_SCRIPT"
}
```
{% include copy-curl.html %}

Like in the bi-encoder mode, use the Tasks API to check the status of the registration task. After the Tasks API returns the task state as `COMPLETED`. Note the `model_id` of the model and the tokenizer you've created; you'll need them for the following steps.

> 

Reviewer 2: @kolchfa-aws See my addition
